shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java

```diff
@@ -191,9 +196,21 @@ private void initializeFlutter(@NonNull Context context) {
   private void setupFlutterEngine() {
     Log.d(TAG, "Setting up FlutterEngine.");
 
-    // First, defer to subclasses for a custom FlutterEngine.
+    // First, check if the host wants to use a cached FlutterEngine.
+    String cachedEngineId = host.getCachedEngineId();
+    if (cachedEngineId != null) {
+      flutterEngine = FlutterEngineCache.getInstance().get(cachedEngineId);
+      isFlutterEngineFromHost = true;
+      if (flutterEngine == null) {
```

Comment:
This is almost like an assert() to me. I do agree that we should validate that a cached engine exists but only in debug mode if it can be detected here. In release mode, perhaps we create an engine?

Reply:
I considered that. The thing is, we'd be assuming the desired app bundle path, dart entrypoint, and initial route. Maybe we guess right, maybe not. If we get to this point and no cached engine exists, we have a fairly serious logic error. Now, if it's possible for this to happen due some kind of timing issue then we'd want to provide a better resolution, but if the dev simply forgot to put an engine in the cache then I don't think we can continue execution.

Reply:
@mehmetf @xster any further thoughts here? I'm open to alternatives if you think we should do something else. I just don't think we can assume a particular engine configuration to auto-create one.

Reply:
If we have to guess all of that then yes, I agree with you. We should fail.